docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,461 @@
+	// +kubebuilder:validation:Minimum:=0
+	// +required
+	Priority int `json:"priority"`
+
+	// dscp marking value for matching pods' traffic.
+	// +kubebuilder:validation:Maximum:=63
+	// +kubebuilder:validation:Minimum:=0
+	DSCP int `json:"dscp"`
+
+	// classifier The classifier on which packets should match
+	// to apply the NetworkQoS Rule.
+	// This field is optional, and in case it is not set the rule is applied
+	// to all egress traffic regardless of the destination.
+	// +optional
+	Classifier Classifier `json:"classifier"`
+
+	// +optional
+	Bandwidth Bandwidth `json:"bandwidth"`
+}
+
+type Classifier struct {
+	// +optional
+	To []Destination `json:"to"`
+
+	// protocol (tcp, udp, sctp) that the traffic must match.
+	// +kubebuilder:validation:Pattern=^TCP|UDP|SCTP$
+	// +optional
+	Protocol string `json:"protocol"`
+
+	// port that the traffic must match
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=65535
+	// +optional
+	Port int32 `json:"port"`
+}
+
+// Bandwidth controls the maximum of rate traffic that can be sent
+// or received on the matching packets.
+type Bandwidth struct {
+	// rate The value of rate limit in kbps. Traffic over the limit
+	// will be dropped.
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=4294967295
+	// +optional
+	Rate uint32 `json:"rate"`
+
+	// burst The value of burst rate limit in kilobits.
+	// This also needs rate to be specified.
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=4294967295
+	// +optional
+	Burst uint32 `json:"burst"`
+}
+
+// Destination describes a peer to apply NetworkQoS configuration for the outgoing traffic.
+// Only certain combinations of fields are allowed.
+type Destination struct {
+	// podSelector is a label selector which selects pods. This field follows standard label
+	// selector semantics; if present but empty, it selects all pods.
+	//
+	// If namespaceSelector is also set, then the NetworkQoS as a whole selects
+	// the pods matching podSelector in the Namespaces selected by NamespaceSelector.
+	// Otherwise it selects the pods matching podSelector in the NetworkQoS's own namespace.
+	// +optional
+	PodSelector *metav1.LabelSelector `json:"podSelector,omitempty" protobuf:"bytes,1,opt,name=podSelector"`
+
+	// namespaceSelector selects namespaces using cluster-scoped labels. This field follows
+	// standard label selector semantics; if present but empty, it selects all namespaces.
+	//
+	// If podSelector is also set, then the NetworkQoS as a whole selects
+	// the pods matching podSelector in the namespaces selected by namespaceSelector.
+	// Otherwise it selects all pods in the namespaces selected by namespaceSelector.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+
+	// ipBlock defines policy on a particular IPBlock. If this field is set then
+	// neither of the other fields can be.
```

Comment:
I think we can be better in the usage of ipBlock. Instead of `neither of the other fields can be`, how about 
```
If this field is set, it functions as a logical OR in the QoS rule, where the destination will match either the (PodSelector, NamespaceSelector) as described above, or the destination address will match the IPBlock.
```

Reply:
@pperiyasamy wdyt? ^^